library/core/src/ptr/non_null.rs

```diff
@@ -552,6 +553,12 @@ impl<T: ?Sized> NonNull<T> {
     #[must_use = "returns a new pointer rather than modifying its argument"]
     #[stable(feature = "non_null_convenience", since = "1.80.0")]
     #[rustc_const_stable(feature = "non_null_convenience", since = "1.80.0")]
+    #[requires(count.checked_mul(core::mem::size_of::<T>()).is_some() 
+        && count * core::mem::size_of::<T>() <= isize::MAX as usize
+        && (self.pointer as isize).checked_add(count as isize * core::mem::size_of::<T>() as isize).is_some() // check wrapping add
+        && kani::mem::same_allocation(self.pointer, self.pointer.wrapping_offset(count as isize)))]
+    #[ensures(|result: &NonNull<T>| result.as_ptr() == self.as_ptr().offset(count as isize)
+        && kani::mem::same_allocation(result.pointer, self.pointer))]
```

Comment:
The second conjunct of the ensures clause (`kani::mem::same_allocation`) is not part of the function's spec.

Reply:
The Safety requirement states: 
If the computed offset is non-zero, then `self` must be derived from a pointer to some [allocated object], and the entire memory range between `self` and the result must be in bounds of that allocated object. Doesn't that mean the resulted pointer should be pointing to the same object as the given self pointer?

Reply:
Yes, but this is a _pre-condition_ of the function, and you already have it in the `requires` clause, so this should be sufficient.